third_party/move/tools/move-resource-viewer/src/lib.rs

```diff
@@ -99,22 +101,28 @@ pub enum AnnotatedMoveValue {
 
 pub struct MoveValueAnnotator<V> {
     module_viewer: V,
+    /// A cache for fat type info for structs. For a generic struct, the uninstantiated
+    /// FatStructType of the base definition will be stored here as well.
+    fat_struct_def_cache: RefCell<BTreeMap<StructName, FatStructRef>>,
+    /// A cache for fat type info for struct instantiations. This cache is build from
+    /// substituting parameters for the uninstantiated types in `fat_struct_def_cache`.
+    fat_struct_inst_cache: RefCell<BTreeMap<(StructName, Vec<FatType>), FatStructRef>>,
+}
+
+#[derive(Debug, Clone, PartialOrd, Ord, PartialEq, Eq, Hash)]
+struct StructName {
+    address: AccountAddress,
+    module: Identifier,
+    name: Identifier,
 }
 
 impl<V: CompiledModuleView> MoveValueAnnotator<V> {
     pub fn new(module_viewer: V) -> Self {
-        Self { module_viewer }
-    }
-
-    pub fn get_type_layout_runtime(&self, type_tag: &TypeTag) -> anyhow::Result<MoveTypeLayout> {
-        TypeLayoutBuilder::build_runtime(type_tag, &self.module_viewer)
-    }
-
-    pub fn get_type_layout_with_fields(
-        &self,
-        type_tag: &TypeTag,
-    ) -> anyhow::Result<MoveTypeLayout> {
-        TypeLayoutBuilder::build_with_fields(type_tag, &self.module_viewer)
+        Self {
+            module_viewer,
+            fat_struct_def_cache: RefCell::default(),
```

Comment:
Once thing that becomes wrong now with caches are limits, `Limiter` may not error when it could have before: this is probably fine as this is not gas and just to bound execution to some fixed number of steps, probably is worth mentioning somewhere?